atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,223 @@
+
+        CommitInfo commitInfo = maybeCommitInfo.get();
+
+        ClientLogEvents update = eventLog.getEventsBetweenVersions(startVersion, commitInfo.commitVersion());
+
+        if (update.clearCache()) {
+            return ImmutableInvalidateAll.builder().build();
+        }
+
+        return constructCommitUpdate(commitInfo, update.events());
+    }
+
+    @Override
+    public TransactionsLockWatchEvents getEventsForTransactions(
+            Set<Long> startTimestamps,
+            Optional<IdentifiedVersion> startVersion) {
+        Preconditions.checkArgument(!startTimestamps.isEmpty(), "Cannot get events for empty set of transactions");
+        Map<Long, IdentifiedVersion> timestampToVersion = getTimestampMappings(startTimestamps);
+        IdentifiedVersion endVersion = Collections.max(timestampToVersion.values(),
+                Comparator.comparingLong(IdentifiedVersion::version));
+        return eventLog.getEventsBetweenVersions(startVersion, endVersion).map(timestampToVersion);
+    }
+
+    @Override
+    public void removeTransactionStateFromCache(long startTimestamp) {
+        timestampMap.remove(startTimestamp);
+    }
+
+    @VisibleForTesting
+    Map<Long, IdentifiedVersion> getTimestampMappings(Set<Long> startTimestamps) {
+        Map<Long, IdentifiedVersion> timestampToVersion = new HashMap<>();
+        startTimestamps.forEach(timestamp -> {
+            Optional<IdentifiedVersion> entry = timestampMap.getStartVersion(timestamp);
+            checkConditionOrThrow(!entry.isPresent(), "start timestamp missing from map");
+            timestampToVersion.put(timestamp, entry.get());
+        });
+        return timestampToVersion;
+    }
+
+    @VisibleForTesting
+    Optional<IdentifiedVersion> getEarliestVersion() {
+        Optional<IdentifiedVersion> currentVersion = eventLog.getLatestKnownVersion();
+        return timestampMap.getEarliestVersion().flatMap(sequence ->
+                currentVersion.map(version -> IdentifiedVersion.of(version.id(), sequence)));
+    }
+
+    private void checkConditionOrThrow(boolean condition, String message) {
+        if (condition) {
+            throw new TransactionLockWatchFailedException(message);
+        }
+    }
+
+    private CommitUpdate constructCommitUpdate(CommitInfo commitInfo, List<LockWatchEvent> events) {
+        LockEventVisitor eventVisitor = new LockEventVisitor(commitInfo.commitLockToken());
+        Set<LockDescriptor> locksTakenOut = new HashSet<>();
+        events.forEach(event -> locksTakenOut.addAll(event.accept(eventVisitor)));
+        return ImmutableInvalidateSome.builder().invalidatedLocks(locksTakenOut).build();
+    }
+
+    private Optional<IdentifiedVersion> processEventLogUpdate(LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> currentVersion = eventLog.getLatestKnownVersion();
+        Optional<IdentifiedVersion> latestVersion = eventLog.processUpdate(update);
+
+        if (!(latestVersion.isPresent()
```

Comment:
there are 4 checks here:

1. is the version at all present? This corresponds to a failure `LockWatchStateUpdate`
2. is the current version (i.e. the one before the update) present? Should only be if we are going from either initialisation or a failure to a snapshot/success (and in which case, we want to reset the timestamp store)
3. did the version change?
4. was it a successful update?

Just thinking this through, I could just as easily return a boolean that states whether it was a successful update or not, and then just ask for the latest version on the next line; since this is all synchronised, this won't cause problems and would handle this.

Reply:
Changed.